File: middleware/windowmanager/windowmanager_test.go
package windowmanager

import (
	"context"
	"testing"
)

func TestNewContextWindowManager(t *testing.T) {
	tests := []struct {
		name      string
		model     ModelSpec
		strategy  CompactionStrategy
		counter   TokenCounter
		expectErr bool
	}{
		{
			name: "valid config uses defaults",
			model: ModelSpec{
				MaxTokens:      4096,
				ReservedTokens: 512,
			},
			strategy:  &LastNCompaction{KeepCount: 10},
			counter:   nil,
			expectErr: false,
		},
		{
			name: "zero max tokens uses default",
			model: ModelSpec{
				MaxTokens:      0,
				ReservedTokens: 100,
			},
			strategy:  &LastNCompaction{},
			counter:   nil,
			expectErr: false,
		},
		{
			name: "negative reserved tokens uses default",
			model: ModelSpec{
				MaxTokens:      4096,
				ReservedTokens: -1,
			},
			strategy:  &LastNCompaction{},
			counter:   nil,
			expectErr: false,
		},
		{
			name: "reserved tokens >= max tokens returns error",
			model: ModelSpec{
				MaxTokens:      4096,
				ReservedTokens: 4096,
			},
			strategy:  &LastNCompaction{},
			counter:   nil,
			expectErr: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			_, err := NewContextWindowManager(tt.model, tt.strategy, tt.counter)
			if (err != nil) != tt.expectErr {
				t.Errorf("NewContextWindowManager() error = %v, expectErr %v", err, tt.expectErr)
			}
		})
	}
}

func TestContextWindowManager_Check(t *testing.T) {
	manager, _ := NewContextWindowManager(
		ModelSpec{MaxTokens: 1000, ReservedTokens: 200},
		&LastNCompaction{},
		nil,
	)

	tests := []struct {
		name           string
		messages       []Message
		expectedTokens int
		expectError    bool
	}{
		{
			name:           "empty messages returns zero tokens",
			messages:       []Message{},
			expectedTokens: 0,
			expectError:    false,
		},
		{
			name: "single user message",
			messages: []Message{
				{Role: "user", Content: "hello"},
			},
			expectError: false,
		},
		{
			name: "multiple messages",
			messages: []Message{
				{Role: "system", Content: "You are helpful."},
				{Role: "user", Content: "Hello"},
				{Role: "assistant", Content: "Hi there!"},
			},
			expectError: false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			status, err := manager.Check(context.Background(), tt.messages)
			if (err != nil) != tt.expectError {
				t.Errorf("Check() error = %v, expectError %v", err, tt.expectError)
			}
			if !tt.expectError && status == nil {
				t.Error("Check() returned nil status")
			}
		})
	}
}

func TestContextWindowManager_ShouldCompact(t *testing.T) {
	manager, _ := NewContextWindowManager(
		ModelSpec{MaxTokens: 1000, ReservedTokens: 200},
		&LastNCompaction{},
		nil,
	)

	smallMessages := []Message{
		{Role: "user", Content: "Hi"},
	}

	largeMessages := make([]Message, 100)
	for i := range largeMessages {
		largeMessages[i] = Message{Role: "user", Content: "This is a test message with some content to fill up space."}
	}

	shouldCompact, err := manager.ShouldCompact(context.Background(), smallMessages)
	if err != nil {
		t.Errorf("ShouldCompact() error = %v", err)
	}
	if shouldCompact {
		t.Error("ShouldCompact() should not compact small messages")
	}

	_, err = manager.ShouldCompact(context.Background(), largeMessages)
	if err != nil {
		t.Errorf("ShouldCompact() error = %v", err)
	}
}

func TestContextWindowManager_Compact(t *testing.T) {
	manager, _ := NewContextWindowManager(
		ModelSpec{MaxTokens: 1000, ReservedTokens: 200},
		&LastNCompaction{KeepCount: 2},
		nil,
	)

	messages := []Message{
		{Role: "system", Content: "System prompt"},
		{Role: "user", Content: "Message 1"},
		{Role: "assistant", Content: "Response 1"},
		{Role: "user", Content: "Message 2"},
		{Role: "assistant", Content: "Response 2"},
	}

	compacted, err := manager.Compact(context.Background(), messages)
	if err != nil {
		t.Errorf("Compact() error = %v", err)
	}
	if len(compacted) == 0 {
		t.Error("Compact() returned empty messages")
	}
}

func TestContextWindowManager_CompactIfNeeded(t *testing.T) {
	manager, _ := NewContextWindowManager(
		ModelSpec{MaxTokens: 1000, ReservedTokens: 200},
		&LastNCompaction{KeepCount: 10},
		nil,
	)

	smallMessages := []Message{
		{Role: "user", Content: "Hello"},
	}

	compacted, didCompact, err := manager.CompactIfNeeded(context.Background(), smallMessages)
	if err != nil {
		t.Errorf("CompactIfNeeded() error = %v", err)
	}
	if didCompact {
		t.Error("CompactIfNeeded() should not compact small messages")
	}
	if len(compacted) != len(smallMessages) {
		t.Errorf("CompactIfNeeded() returned %d messages, want %d", len(compacted), len(smallMessages))
	}
}

func TestContextWindowManager_calculateWarningLevel(t *testing.T) {
	manager, _ := NewContextWindowManager(
		ModelSpec{MaxTokens: 1000, ReservedTokens: 0},
		&LastNCompaction{},
		nil,
	)

	tests := []struct {
		name          string
		remaining     int
		available     int
		expectedLevel WarningLevel
	}{
		{"critical 10%", 100, 1000, WarningLevelCritical},
		{"high 25%", 250, 1000, WarningLevelHigh},
		{"medium 50%", 500, 1000, WarningLevelMedium},
		{"low 75%", 750, 1000, WarningLevelLow},
		{"none 90%", 900, 1000, WarningLevelNone},
		{"zero available", 0, 0, WarningLevelCritical},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			level := manager.calculateWarningLevel(tt.remaining, tt.available)
			if level != tt.expectedLevel {
				t.Errorf("calculateWarningLevel() = %v, want %v", level, tt.expectedLevel)
			}
		})
	}
}

func TestLastNCompaction_Compact(t *testing.T) {
	strategy := &LastNCompaction{KeepCount: 2}
	counter := &DefaultCounter{}

	messages := []Message{
		{Role: "system", Content: "System"},
		{Role: "user", Content: "User 1"},
		{Role: "assistant", Content: "Assistant 1"},
		{Role: "user", Content: "User 2"},
		{Role: "assistant", Content: "Assistant 2"},
	}

	compacted, err := strategy.Compact(messages, 100, counter)
	if err != nil {
		t.Errorf("Compact() error = %v", err)
	}
	if len(compacted) > len(messages) {
		t.Errorf("Compact() returned more messages than input: %d > %d", len(compacted), len(messages))
	}
}

func TestLastNCompaction_CompactZeroKeepCount(t *testing.T) {
	strategy := &LastNCompaction{KeepCount: 0}
	counter := &DefaultCounter{}

	messages := []Message{
		{Role: "user", Content: "Message 1"},
		{Role: "user", Content: "Message 2"},
	}

	compacted, err := strategy.Compact(messages, 1000, counter)
	if err != nil {
		t.Errorf("Compact() error = %v", err)
	}
	if len(compacted) == 0 {
		t.Error("Compact() should return at least one message")
	}
}

func TestLastNCompaction_CompactEmpty(t *testing.T) {
	strategy := &LastNCompaction{}
	counter := &DefaultCounter{}

	compacted, err := strategy.Compact([]Message{}, 100, counter)
	if err != nil {
		t.Errorf("Compact() error = %v", err)
	}
	if len(compacted) != 0 {
		t.Error("Compact() should return empty for empty input")
	}
}

func TestSummaryCompaction_Compact(t *testing.T) {
	strategy := NewSummaryCompaction()
	counter := &DefaultCounter{}

	messages := []Message{
		{Role: "system", Content: "System prompt here"},
		{Role: "user", Content: "User message 1"},
		{Role: "assistant", Content: "Assistant response 1"},
		{Role: "user", Content: "User message 2"},
		{Role: "assistant", Content: "Assistant response 2"},
	}

	compacted, err := strategy.Compact(messages, 100, counter)
	if err != nil {
		t.Errorf("Compact() error = %v", err)
	}
	if len(compacted) == 0 {
		t.Error("Compact() should return messages")
	}
}

func TestSummaryCompaction_CompactEmpty(t *testing.T) {
	strategy := NewSummaryCompaction()
	counter := &DefaultCounter{}

	compacted, err := strategy.Compact([]Message{}, 100, counter)
	if err != nil {
		t.Errorf("Compact() error = %v", err)
	}
	if len(compacted) != 0 {
		t.Error("Compact() should return empty for empty input")
	}
}

func TestSummaryCompaction_summarizeMessages(t *testing.T) {
	strategy := NewSummaryCompaction()

	messages := []Message{
		{Role: "user", Content: "Hello"},
		{Role: "assistant", Content: "Hi there"},
	}

	summary := strategy.summarizeMessages(messages, 100)
	if summary == "" {
		t.Error("summarizeMessages() should return non-empty string")
	}

	shortSummary := strategy.summarizeMessages(messages, 5)
	if len(shortSummary) > 5 {
		t.Errorf("summarizeMessages() should truncate to maxLen, got %d", len(shortSummary))
	}
}

func TestPriorityBasedCompaction_Compact(t *testing.T) {
	strategy := NewPriorityBasedCompaction()
	counter := &DefaultCounter{}

	messages := []Message{
		{Role: "system", Content: "System"},
		{Role: "user", Content: "User"},
		{Role: "assistant", Content: "Assistant"},
	}

	compacted, err := strategy.Compact(messages, 1000, counter)
	if err != nil {
		t.Errorf("Compact() error = %v", err)
	}
	if len(compacted) == 0 {
		t.Error("Compact() should return messages")
	}
}

func TestPriorityBasedCompaction_CompactEmpty(t *testing.T) {
	strategy := NewPriorityBasedCompaction()
	counter := &DefaultCounter{}

	compacted, err := strategy.Compact([]Message{}, 100, counter)
	if err != nil {
		t.Errorf("Compact() error = %v", err)
	}
	if len(compacted) != 0 {
		t.Error("Compact() should return empty for empty input")
	}
}

func TestPriorityBasedCompaction_Name(t *testing.T) {
	strategy := NewPriorityBasedCompaction()
	if strategy.Name() != "priority" {
		t.Errorf("Name() = %v, want priority", strategy.Name())
	}
}

func TestDefaultCounter_Count(t *testing.T) {
	counter := &DefaultCounter{}

	messages := []Message{
		{Role: "user", Content: "Hello"},
		{Role: "assistant", Content: "Hi there"},
	}

	count, err := counter.Count(messages)
	if err != nil {
		t.Errorf("Count() error = %v", err)
	}
	if count <= 0 {
		t.Error("Count() should return positive count")
	}
}

func TestDefaultCounter_CountMessage(t *testing.T) {
	counter := &DefaultCounter{}

	msg := Message{Role: "user", Content: "Test message", Name: "User1"}

	count, err := counter.CountMessage(msg)
	if err != nil {
		t.Errorf("CountMessage() error = %v", err)
	}
	if count <= 0 {
		t.Error("CountMessage() should return positive count")
	}
}

func TestDefaultCounter_CountEmptyMessage(t *testing.T) {
	counter := &DefaultCounter{}

	msg := Message{Role: "", Content: ""}

	count, err := counter.CountMessage(msg)
	if err != nil {
		t.Errorf("CountMessage() error = %v", err)
	}
	if count < 4 {
		t.Errorf("CountMessage() should return at least 4 for overhead, got %d", count)
	}
}

func TestWarningLevel_Constants(t *testing.T) {
	if WarningLevelNone != "none" {
		t.Errorf("WarningLevelNone = %v, want none", WarningLevelNone)
	}
	if WarningLevelLow != "low" {
		t.Errorf("WarningLevelLow = %v, want low", WarningLevelLow)
	}
	if WarningLevelMedium != "medium" {
		t.Errorf("WarningLevelMedium = %v, want medium", WarningLevelMedium)
	}
	if WarningLevelHigh != "high" {
		t.Errorf("WarningLevelHigh = %v, want high", WarningLevelHigh)
	}
	if WarningLevelCritical != "critical" {
		t.Errorf("WarningLevelCritical = %v, want critical", WarningLevelCritical)
	}
}

func TestLastNCompaction_Name(t *testing.T) {
	strategy := &LastNCompaction{KeepCount: 5}
	if strategy.Name() != "last_n" {
		t.Errorf("Name() = %v, want last_n", strategy.Name())
	}
}

func TestSummaryCompaction_Name(t *testing.T) {
	strategy := NewSummaryCompaction()
	if strategy.Name() != "summary" {
		t.Errorf("Name() = %v, want summary", strategy.Name())
	}
}

func TestCompactionStrategy_Interface(t *testing.T) {
	var _ CompactionStrategy = (*LastNCompaction)(nil)
	var _ CompactionStrategy = (*SummaryCompaction)(nil)
	var _ CompactionStrategy = (*PriorityBasedCompaction)(nil)
}

func TestTokenCounter_Interface(t *testing.T) {
	var _ TokenCounter = (*DefaultCounter)(nil)
}

File: middleware_py/tests/windowmanager_test.py
"""Tests for middleware_py.windowmanager."""

import pytest

from middleware_py.windowmanager import (
    WarningLevel,
    Message,
    ModelSpec,
    ContextStatus,
    CompactionStrategy,
    TokenCounter,
    ContextWindowManager,
    LastNCompaction,
    Priority,
    PriorityMessage,
    PriorityBasedCompaction,
    SummaryCompaction,
    DefaultCounter,
)


class TestModelSpec:
    def test_default_values(self):
        spec = ModelSpec()
        assert spec.name == ""
        assert spec.max_tokens == 4096
        assert spec.reserved_tokens == 0
        assert spec.token_multiplier == 4.0

    def test_custom_values(self):
        spec = ModelSpec(name="gpt-4", max_tokens=8192, reserved_tokens=1024)
        assert spec.name == "gpt-4"
        assert spec.max_tokens == 8192
        assert spec.reserved_tokens == 1024


class TestMessage:
    def test_basic_message(self):
        msg = Message(role="user", content="Hello")
        assert msg.role == "user"
        assert msg.content == "Hello"
        assert msg.name == ""

    def test_message_with_name(self):
        msg = Message(role="user", content="Hello", name="user1")
        assert msg.name == "user1"


class TestWarningLevel:
    def test_warning_levels(self):
        assert WarningLevel.NONE.value == "none"
        assert WarningLevel.LOW.value == "low"
        assert WarningLevel.MEDIUM.value == "medium"
        assert WarningLevel.HIGH.value == "high"
        assert WarningLevel.CRITICAL.value == "critical"


class TestContextWindowManager:
    def test_init_with_defaults(self):
        manager = ContextWindowManager(ModelSpec())
        assert manager._model.max_tokens == 4096
        assert isinstance(manager._strategy, LastNCompaction)
        assert isinstance(manager._counter, DefaultCounter)

    def test_init_invalid_max_tokens(self):
        model = ModelSpec(max_tokens=0)
        manager = ContextWindowManager(model)
        assert manager._model.max_tokens == 4096

    def test_init_invalid_reserved_tokens(self):
        model = ModelSpec(reserved_tokens=-1)
        manager = ContextWindowManager(model)
        assert manager._model.reserved_tokens == 0

    def test_init_reserved_exceeds_max(self):
        model = ModelSpec(max_tokens=1000, reserved_tokens=2000)
        with pytest.raises(ValueError, match="reserved_tokens must be less than max_tokens"):
            ContextWindowManager(model)

    def test_init_zero_multiplier(self):
        model = ModelSpec(token_multiplier=0)
        manager = ContextWindowManager(model)
        assert manager._model.token_multiplier == 4.0

    def test_check_empty_messages(self):
        manager = ContextWindowManager(ModelSpec(max_tokens=4096, reserved_tokens=512))
        status = manager.check([])
        assert status.used_tokens == 0
        assert status.remaining_tokens == 4096 - 512
        assert status.warning_level == WarningLevel.NONE
        assert status.message_count == 0

    def test_check_with_messages(self):
        manager = ContextWindowManager(ModelSpec(max_tokens=4096))
        messages = [
            Message(role="user", content="Hello world this is a test"),
        ]
        status = manager.check(messages)
        assert status.used_tokens > 0
        assert status.message_count == 1

    def test_check_remaining_tokens(self):
        manager = ContextWindowManager(ModelSpec(max_tokens=1000, reserved_tokens=200))
        messages = [Message(role="user", content="a" * 100)]
        status = manager.check(messages)
        assert status.remaining_tokens == 800 - status.used_tokens

    def test_warning_level_calculation_critical(self):
        manager = ContextWindowManager(ModelSpec(max_tokens=1000, reserved_tokens=0))
        status = manager.check([Message(role="user", content="a" * 3584)])
        assert status.warning_level == WarningLevel.CRITICAL

    def test_warning_level_calculation_high(self):
        manager = ContextWindowManager(ModelSpec(max_tokens=1000, reserved_tokens=0))
        status = manager.check([Message(role="user", content="a" * 2984)])
        assert status.warning_level == WarningLevel.HIGH

    def test_warning_level_calculation_medium(self):
        manager = ContextWindowManager(ModelSpec(max_tokens=1000, reserved_tokens=0))
        status = manager.check([Message(role="user", content="a" * 1984)])
        assert status.warning_level == WarningLevel.MEDIUM

    def test_warning_level_calculation_low(self):
        manager = ContextWindowManager(ModelSpec(max_tokens=1000, reserved_tokens=0))
        status = manager.check([Message(role="user", content="a" * 984)])
        assert status.warning_level == WarningLevel.LOW

    def test_warning_level_calculation_none(self):
        manager = ContextWindowManager(ModelSpec(max_tokens=1000, reserved_tokens=0))
        status = manager.check([Message(role="user", content="a" * 50)])
        assert status.warning_level == WarningLevel.NONE

    def test_should_compact_true(self):
        manager = ContextWindowManager(ModelSpec(max_tokens=1000))
        messages = [Message(role="user", content="a" * 3584)]
        assert manager.should_compact(messages) is True

    def test_should_compact_false(self):
        manager = ContextWindowManager(ModelSpec(max_tokens=1000))
        messages = [Message(role="user", content="a" * 100)]
        assert manager.should_compact(messages) is False

    def test_compact_empty_messages(self):
        manager = ContextWindowManager(ModelSpec())
        result = manager.compact([])
        assert result == []

    def test_compact_needed(self):
        manager = ContextWindowManager(ModelSpec(max_tokens=1000))
        messages = [Message(role="user", content="a" * 4001)]
        result = manager.compact(messages)
        assert len(result) <= len(messages)

    def test_compact_if_needed_no_compact(self):
        manager = ContextWindowManager(ModelSpec(max_tokens=1000))
        messages = [Message(role="user", content="a" * 100)]
        result, did_compact = manager.compact_if_needed(messages)
        assert result == messages
        assert did_compact is False

    def test_compact_if_needed_with_compact(self):
        manager = ContextWindowManager(ModelSpec(max_tokens=1000))
        messages = [Message(role="user", content="a" * 4001)]
        result, did_compact = manager.compact_if_needed(messages)
        assert did_compact is True
        assert len(result) <= len(messages)


class TestLastNCompaction:
    def test_default_keep_count(self):
        strategy = LastNCompaction()
        messages = [
            Message(role="user", content="a"),
            Message(role="assistant", content="b"),
            Message(role="user", content="c"),
            Message(role="assistant", content="d"),
        ]
        counter = DefaultCounter()
        result = strategy.compact(messages, 11, counter)
        assert len(result) == 2

    def test_custom_keep_count(self):
        strategy = LastNCompaction(keep_count=1)
        messages = [
            Message(role="user", content="a"),
            Message(role="assistant", content="b"),
        ]
        counter = DefaultCounter()
        result = strategy.compact(messages, 100, counter)
        assert len(result) == 1

    def test_compact_empty_messages(self):
        strategy = LastNCompaction()
        counter = DefaultCounter()
        result = strategy.compact([], 100, counter)
        assert result == []

    def test_compact_already_fits(self):
        strategy = LastNCompaction()
        messages = [Message(role="user", content="a")]
        counter = DefaultCounter()
        result = strategy.compact(messages, 100, counter)
        assert result == messages

    def test_compact_returns_last_message(self):
        strategy = LastNCompaction(keep_count=0)
        messages = [
            Message(role="user", content="a"),
            Message(role="assistant", content="b"),
            Message(role="user", content="c"),
        ]
        counter = DefaultCounter()
        result = strategy.compact(messages, 1, counter)
        assert len(result) == 1


class TestPriorityBasedCompaction:
    def test_name(self):
        strategy = PriorityBasedCompaction()
        assert strategy.name() == "priority"

    def test_keep_system_by_default(self):
        strategy = PriorityBasedCompaction(keep_system=True)
        messages = [
            Message(role="system", content="system prompt"),
            Message(role="user", content="user message"),
            Message(role="assistant", content="assistant message"),
        ]
        counter = DefaultCounter()
        result = strategy.compact(messages, 100, counter)
        assert result[0].role == "system"

    def test_compact_empty(self):
        strategy = PriorityBasedCompaction()
        counter = DefaultCounter()
        result = strategy.compact([], 100, counter)
        assert result == []

    def test_priority_order(self):
        strategy = PriorityBasedCompaction(keep_system=False)
        messages = [
            Message(role="system", content="system"),
            Message(role="user", content="user"),
            Message(role="assistant", content="assistant"),
            Message(role="unknown", content="unknown"),
        ]
        counter = DefaultCounter()
        result = strategy.compact(messages, 100, counter)
        assert result[0].role == "assistant"


class TestSummaryCompaction:
    def test_name(self):
        strategy = SummaryCompaction()
        assert strategy.name() == "summary"

    def test_compact_empty(self):
        strategy = SummaryCompaction()
        counter = DefaultCounter()
        result = strategy.compact([], 100, counter)
        assert result == []

    def test_no_system_message(self):
        strategy = SummaryCompaction()
        messages = [
            Message(role="user", content="hello"),
            Message(role="assistant", content="hi there"),
        ]
        counter = DefaultCounter()
        result = strategy.compact(messages, 100, counter)
        assert len(result) > 0

    def test_with_system_message(self):
        strategy = SummaryCompaction()
        messages = [
            Message(role="system", content="You are helpful."),
            Message(role="user", content="hello"),
            Message(role="assistant", content="hi there"),
        ]
        counter = DefaultCounter()
        result = strategy.compact(messages, 100, counter)
        assert result[0].role == "system"

    def test_summary_contains_prefix(self):
        strategy = SummaryCompaction()
        messages = [
            Message(role="user", content="test"),
            Message(role="assistant", content="result"),
        ]
        counter = DefaultCounter()
        result = strategy.compact(messages, 1, counter)
        assert len(result) > 0


class TestDefaultCounter:
    def test_count_empty(self):
        counter = DefaultCounter()
        assert counter.count([]) == 0

    def test_count_single_message(self):
        counter = DefaultCounter()
        msg = Message(role="user", content="hello world")
        count = counter.count_message(msg)
        assert count > 0

    def test_count_multiple_messages(self):
        counter = DefaultCounter()
        messages = [
            Message(role="user", content="hello"),
            Message(role="assistant", content="world"),
        ]
        total = counter.count(messages)
        assert total > 0

    def test_count_message_with_name(self):
        counter = DefaultCounter()
        msg = Message(role="user", content="hello", name="user1")
        count = counter.count_message(msg)
        assert count > 0